teps/000X-trigger-crd.md

```diff
@@ -0,0 +1,370 @@
+-->
+Today, EventListener commonly are created for every namespace and handle TriggerBinding and
+TriggerTemplate in that same namespace. Creation of EventListener causes pod proliferation. For every 
+namespace that requires handling of webhook events like github, we need EventListener which in turns 
+lead to this pod proliferation and causes excess resource consumption. In multitenant scenario where
+projects are divided according to namespaces, we require EventListener for each namespaces. So for 
+100 of projects, we would need 100 of pods just for acting as sink for these webhook events like github.
+These pods in turn consume resources.
+
+### Goals
+
+<!--
+List the specific goals of the TEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+1. Reduction in resource cosumption due to EventListener.
+2. Ability to have single EventListener to cater to whole cluster.
+3. Separation of Concerns :  The more ELs there are, the more additional configuration are needed - in some cases this additional bit of configuration is something that a cluster admin/operator would perform not an user/app developer.
+For declarative config as code scenarios, it would be easier for a user to have create the Trigger configuration part as part of config with code. The cluster admin/operator can create the EventListeners, expose it to a public address, etc. And the same listener can be shared among many Triggers. This does reduce the number of listener Pods but also allows for a better separation of concerns among the user and the operator.
+
+### Non-Goals
+
+<!--
+What is out of scope for this TEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+How to create Multitenant EventListener resource with appropriate permissions won't be part of this proposal.
+This will be handled by operator or admin of the cluster.
+
+## Requirements
+
+<!--
+List the requirements for this TEP.
+-->
+1. Can create Tekton resources in a single namespace or across multiple namespaces if configured to do so.
+2. When an EL recieves an event, it can process only some configured subset of triggers <- path/selector based
+3. EL can still process triggers defined in the current form aka backwards compatibility.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+EventListener CRD will be split into two - a Trigger CRD and a EventListener CRD.
+
+### Trigger CRD
+A single Trigger defines configuration for processing events i.e. it consists of TriggerBindings,
+a single TriggerTemplate, and optionally interceptors and a serviceAccount. 
+```
+apiVersion: v1alpha1
+kind: Trigger
+metadata:
+  name: my-repo-trigger
+spec:
+  serviceAccountName: "blah"
```

Comment:
Observation:  specifying just the SA name here, with the implication that the SA is in the same namespace as the `Trigger` is a change from https://github.com/tektoncd/triggers/blob/master/pkg/apis/triggers/v1alpha1/trigger_types.go#L40-L42 where a fully object ref, including namespace was provided.

Now, from a "security" perspective, this is a positive move IMO.

That said, in our prior discussions on this, I am not sure if we nailed down the existing scenarios that required a full object reference there.

Do any of the reviewers recall?

From a TEP perspective, calling out that difference is probably important.

I still somewhat wonder if independent of this TEP, we should switch that ref from global to namespace scope ...

Reply:
The implication being that a SA from a privileged namespace could be specified with a ObjectReference vs. a LocalObjectReference.

Reply:
> That said, in our prior discussions on this, I am not sure if we nailed down the existing scenarios that required a full object reference there.

I think you are correct. And I agree we should call this out. We should discuss this - this has come up in https://github.com/tektoncd/triggers/issues/679 maybe we can continue the discussion there?